### ISSUE TITLE
BSOD when injecting on Windows 10 protected by KPTI

### ISSUE THREAD

**Issue by mtarral** [tklengyel/drakvuf#622]
Hi,

I would like to report a BSOD issue that i'm facing on Windows 10.

The context is particular, as i can only trigger if the following conditions are met:
- kernel has `Meltdown` protection (`KPTI`)
- process injection is used, simple syscall monitoring doesn't trigger the bug so far

It fails really early in my test suite, often after the 4th try:

![7d06d77188164f75bc774e2c6f7d13ac](https://user-images.githubusercontent.com/49681011/56747010-89ae8680-677d-11e9-84cd-71ba29c30dd7.png)

## BSOD

The `BSOD` message is an `UNEXPECTED_KERNEL_MODE_TRAP`

![8af800dd324d4d8f86a00e8a2dc69b1c](https://user-images.githubusercontent.com/49681011/56747028-8fa46780-677d-11e9-966b-0573de678c33.png)

## WinDBG

Looking closer at the bug with `WinDBG`:

~~~
kd> !analyze -v
*******************************************************************************
*                                                                             *
*                        Bugcheck Analysis                                    *
*                                                                             *
*******************************************************************************

UNEXPECTED_KERNEL_MODE_TRAP (7f)
This means a trap occurred in kernel mode, and it's a trap of a kind
that the kernel isn't allowed to have/catch (bound trap) or that
is always instant death (double fault).  The first number in the
bugcheck params is the number of the trap (8 = double fault, etc)
Consult an Intel x86 family manual to learn more about what these
traps are. Here is a *portion* of those codes:
If kv shows a taskGate
        use .tss on the part before the colon, then kv.
Else if kv shows a trapframe
        use .trap on that value
Else
        .trap on the appropriate frame will show where the trap was taken
        (on x86, this will be the ebp that goes with the procedure KiTrap)
Endif
kb will then show the corrected stack.
Arguments:
Arg1: 0000000000000008, EXCEPTION_DOUBLE_FAULT
Arg2: fffff80354c6ce50
Arg3: fffffb837eea7c90
Arg4: fffff80352b7c12d

...
PROCESS_NAME:  ctfmon.exe
~~~

`TRAP_FRAME`
~~~
TRAP_FRAME:  fffff80354c6ce50 -- (.trap 0xfffff80354c6ce50)
NOTE: The trap frame does not contain all registers.
Some register values may be zeroed or incorrect.
rax=00000000000013ea rbx=0000000000000000 rcx=00007ffb2a468d64
rdx=000001fc1525cd20 rsi=0000000000000000 rdi=0000000000000000
rip=fffff80352b7c12d rsp=fffffb837eea7c90 rbp=000000a73f3ff878
 r8=0000000000000000  r9=0000000000000000 r10=0000000000000008
r11=0000000000000246 r12=0000000000000000 r13=0000000000000000
r14=0000000000000000 r15=0000000000000000
iopl=0         nv up di pl zr na po nc
nt!KiSystemCall64ShadowCommon:
fffff803`52b7c12d 6a2b            push    2Bh
Resetting default scope
~~~

`FOLLOWUP_IP`
~~~
FOLLOWUP_IP: 
nt!KiDoubleFaultAbort+32c
fffff803`52a4322c 90              nop
~~~

`Call stack`
~~~
kd> kn
 # Child-SP          RetAddr           Call Site
00 fffff803`54c6cd08 fffff803`52a48529 nt!KeBugCheckEx
01 fffff803`54c6cd10 fffff803`52a4322c nt!KiBugCheckDispatch+0x69
02 fffff803`54c6ce50 fffff803`52b7c12d nt!KiDoubleFaultAbort+0x32c
03 fffffb83`7eea7c90 fffffb83`7eea8000 nt!KiSystemCall64ShadowCommon
04 fffffb83`7eea7c98 fffffb83`7eea2000 0xfffffb83`7eea8000
05 fffffb83`7eea7ca0 00000000`00000000 0xfffffb83`7eea2000
~~~

`registers`
~~~
kd> kn
 # Child-SP          RetAddr           Call Site
00 fffff803`54c6cd08 fffff803`52a48529 nt!KeBugCheckEx
01 fffff803`54c6cd10 fffff803`52a4322c nt!KiBugCheckDispatch+0x69
02 fffff803`54c6ce50 fffff803`52b7c12d nt!KiDoubleFaultAbort+0x32c
03 fffffb83`7eea7c90 fffffb83`7eea8000 nt!KiSystemCall64ShadowCommon
04 fffffb83`7eea7c98 fffffb83`7eea2000 0xfffffb83`7eea8000
05 fffffb83`7eea7ca0 00000000`00000000 0xfffffb83`7eea2000
kd> r
rax=fffff80354c6ce10 rbx=000001fc1525cd20 rcx=000000000000007f
rdx=0000000000000008 rsi=0000000000000000 rdi=0000000000000000
rip=fffff80352a35650 rsp=fffff80354c6cd08 rbp=fffff80354c6ced0
 r8=fffff80354c6ce50  r9=fffffb837eea7c90 r10=fffff80352b7c12d
r11=0000000000000246 r12=0000000000000000 r13=000001fc1382bda0
r14=0000000000000001 r15=000001fc1385d240
iopl=0         nv up di ng nz na pe nc
cs=0010  ss=0018  ds=002b  es=002b  fs=0053  gs=002b             efl=00000082
nt!KeBugCheckEx:
fffff803`52a35650 48894c2408      mov     qword ptr [rsp+8],rcx ss:0018:fffff803`54c6cd10=000000000000007f
~~~

## Analysis

We are facing a `DoubleFault` exception generated at
~~~
nt!KiSystemCall64ShadowCommon:
fffff803`52b7c12d 6a2b            push    2Bh
~~~

Looking at this function in `radare2`:

![a056e35d4c5644529f1f1bd33738c571](https://user-images.githubusercontent.com/49681011/56747048-97fca280-677d-11e9-9cd4-a080f3b3e54d.png)

`KiSystemCall64Shadow` is responsible for switching to the kernel page tables if we are running on a vulnerable CPU.

Interestingly, the bug appears on the first instruction of `KiSystemCall64ShadowCommon`, after the page table switch.

Trying to see if this virtual address was in physical memory:
~~~
kd> dt _EPROCESS @$proc ImageFileName Pcb.DirectoryTableBase Pcb.UserDirectoryTableBase
ntdll!_EPROCESS
   +0x000 Pcb                        : 
      +0x028 DirectoryTableBase         : 0xcdf00002
      +0x278 UserDirectoryTableBase     : 0xcde00001
   +0x450 ImageFileName              : [15]  "ctfmon.exe"
~~~

`vtop DirectoryTableBase`
~~~
kd> !vtop 0xcdf00002 nt!KiSystemCall64ShadowCommon
Amd64VtoP: Virt 0000000000000000, pagedir 00000000cdf00002
Amd64VtoP: PML4E 00000000cdf00002
Amd64VtoP: pagefile PDPE 0:00000d8670a00000
Amd64VtoP: PML4E not present, 0xD0000147
Virtual address 0 translation fails, error 0xD0000147.
~~~

`vtop UserDirectoryTableBase` (to be sure)
~~~
kd> !vtop 0xcde00001 nt!KiSystemCall64ShadowCommon
Amd64VtoP: Virt 0000000000000000, pagedir 00000000cde00001
Amd64VtoP: PML4E 00000000cde00001
Amd64VtoP: pagefile PDPE 0:00000670a0000000
Amd64VtoP: PML4E not present, 0xD0000147
Virtual address 0 translation fails, error 0xD0000147.
~~~

So it looks like that even after the page table switch, we are still paged-out.

What do you think ?

(Please note that i'm fairly new to `WinDBG`, so there might be mistakes in interpreting the commands output)

Running on Xen 4.12.0

Thanks !

**Comment by tklengyel** [comment]
What mode of the injector are you using?

**Comment by mtarral** [comment]
I have tried with `createproc` and `shellexec`.
Both of them trigger the bug.

**Comment by mtarral** [comment]
I updated my test suite to inject the process via `Ansible` and `WinRM` as guest agent.

The result is the same, Windows 10 is crashing at some point.

Now we can exclude that `DRAKVUF`'s code injection could have, at some point, corrupted the kernel memory and the pagetables.

Which leaves us with the question of how `DRAKVUF`'s monitoring via `altp2m` and `EPT` manipulation can result in the guest having a pagefault ?

And pagefaults are just one example, because I have crashdumps of userland application crashing on very stranges errors, like `invalid opcode`, or `access violation` on a `ret` instruction, trying to access the stack:

# App crashes

## Unknown software exception
![unknow_software_exception](https://user-images.githubusercontent.com/49681011/57142099-06fc7b80-6dbc-11e9-8961-1c8b02c30ce6.png)

## Invalid opcode
![unthauthrized_instruction](https://user-images.githubusercontent.com/49681011/57142113-0ebc2000-6dbc-11e9-9e63-4b964721d23a.png)

## Memory cannot be written
![not_written](https://user-images.githubusercontent.com/49681011/57142142-1e3b6900-6dbc-11e9-91a8-3790e8ac53f1.png)


-> Is it a bug in the way `DRAKVUF` is using the `altp2m` API (forgetting to flush caches ?)
-> is it a rare race condition in Xen's `altp2m` API ?

I'm ready to investigate this @tklengyel , but I don't know how to.
The way the guest is crashing and what could have triggered it is still unclear to me.

I would to send a message on `xen-devel` as well to get more opinions about this issue.

Thanks.

**Comment by tklengyel** [comment]
> -> Is it a bug in the way DRAKVUF is using the altp2m API (forgetting to flush caches ?)

Switching altp2ms assigns a new VPID to the vCPU thus effectively schedules it with a clean cacheline.

>  -> is it a rare race condition in Xen's altp2m API ?

There is no race, the tables don't change after they are setup - and when they change the entire domain is paused. The only time altp2m switches occur is when the vCPU is paused as we are in the callback of an event.

**Comment by tklengyel** [comment]
> I'm ready to investigate this @tklengyel , but I don't know how to.
> The way the guest is crashing and what could have triggered it is still unclear to me.

That's the problem. Without understanding the root cause of the crashes we stand little chance in figuring this out. Obviously there is a race-condition somewhere since this only triggers occasionally. It also doesn't trigger with all plugins - syscall tracing, etc - all work without a problem. So to me that rules out the problem being in lower-layer functions such as Xen's altp2m or the shadow-page implementation of DRAKVUF.

Function-call injection is the most intrusive aspect of DRAKVUF as that visible alters the execution of the guest and modifies its memory. If that's the only time the problem is observed - and only with Windows 10 - then likely there is some additional steps that need to be implemented there to make it safe. Could be there is some type of APC pending for the process we hijack, pushing the control-flow into an unexpected system state.

**Comment by mtarral** [comment]
> It also doesn't trigger with all plugins - syscall tracing, etc - all work without a problem.

Regarding the application crashes, i can't exactly remember if i had all plugins activated, or just the syscalls.

But regarding the BSOD, my test suite invokes drakvuf by activating only `procmon` (to trace guest process until it finishes it execution), `bsodmon` and `crashmon` (to detect a crash and mark the test as failed).
Procmon only monitors a few syscalls, and bsodmon and crashmon are even smaller.

> Function-call injection is the most intrusive aspect of DRAKVUF as that visible alters the execution of the guest and modifies its memory. If that's the only time the problem is observed ...

We have excluded the function call injection issue by replicating the bug with `Ansible` and `WinRM`.

**Comment by tklengyel** [comment]
> We have excluded the function call injection issue by replicating the bug with Ansible and WinRM.

You mean the crash occurs when you don't use injection at all?

Also, try to see if having the guest running with a single vCPU still triggers the crash or not.

**Comment by mtarral** [comment]
> You mean the crash occurs when you don't use injection at all?

Yes, i should have removed the term `injection` here, but i'm only invoking `ansible` to execute a process in the guest.
`ansible '*' -i '<ip>,' -c winrm ... -m win_command -a c:\Windows\System32\reg.exe`

> Also, try to see if having the guest running with a single vCPU still triggers the crash or not.

Yes, Windows 10 is running with 1 VCPU, and it crashes quite fast (20 tests)

**Comment by tklengyel** [comment]
Can you give me your CPU info and the Xen command-line you use to boot?

**Comment by mtarral** [comment]
> Can you give me your CPU info and the Xen command-line you use to boot?

CPU: `Intel(R) Core(TM) i7-4790`
Xen boot command line:
`GRUB_CMDLINE_XEN_DEFAULT="console=tty vt.handoff=7 loglvl=all guest_loglvl=all dom0_mem=7000M,max:7000M dom0_max_vcpus=4 dom0_vcpus_pin=true hap_1gb=false hap_2mb=false altp2m=1"`

**Comment by tklengyel** [comment]
Thanks, your CPU is old enough that it doesn't have `PML` so the Xen bug I was curious to see if it could affect you doesn't apply.

**Comment by tklengyel** [comment]
You could perhaps try the same stuff that seemed to have helped in #621 to see if it makes the bug go away as well.

**Comment by mtarral** [comment]
So, I disabled Spectre/Meltdown protections via [InSpectre](https://www.grc.com/inspectre.htm), and i had to update my `libvmi.conf` as expected.

The results are the following:
I tested `100000` guest process start with Ansible, with **1 VCPU**, and the whole thing remained stable.
(I tested for more than 24 hours, everything was fine, so i declared it stable and closed the tests)

![win10_ansible_100000](https://user-images.githubusercontent.com/49681011/57230122-eaf51600-7017-11e9-8ad7-ebbc467b1777.png)

Then I tested in with **2 VCPU**, and it failed with an appcrash:
![win_10_appcrash_2vcpu_sans_kpti](https://user-images.githubusercontent.com/49681011/57230182-0a8c3e80-7018-11e9-8c51-b26def1460d4.png)

or a BSOD when i retried the test:
![Screenshot_20190506_145124](https://user-images.githubusercontent.com/49681011/57230214-1bd54b00-7018-11e9-805d-c75652364ce6.png)


Conclusion for me is that DRAKVUF monitoring under a Windows 10 with KPTI makes it crash really fast, **so KPTI is clearly linked**.

But apart from that, there is **another rare race-condition bug**, probably the exact same that what we discovered on #576.

@skvl , if you have a heavy use of DRAKVUF, did you ever had this type of mysterious issues (appcrash/bsod) ?
@tklengyel any ideas what should we do to understand the KPTI bug ?

Thanks

**Comment by tklengyel** [comment]
I would suggest to start experimenting with the xen-access test tool included in the Xen source tree. You want to pinpoint which event and/or what type of use of altp2m triggers the bug. You effectively want to cut out as many layers as possible to reduce the complexity and be able to pinpoint the issue.

**Comment by aoshiken** [comment]
Do you have the same/similar symptoms when running on a Win7 x64 platform?

I noticed application crashes from time to time and BSoDs with the message "A clock interrupt was not received on a secondary processor within the allocated time interval".

I'm afraid isn't related to this issue but supposed is good to know.

**Comment by skvl** [comment]
@mtarral I don't usually make long runs. But I have seen application crashes several times while running Drakvuf. Without it a VM could run for hours or even days fine.

**Comment by mtarral** [comment]
@tklengyel 
I will check `xen-access`, reducing the layers seems like  good idea, if it doesn't take too mch time.

@aoshiken 
Yes, i get similar results, applications crashing, but most of all BSODs (`PAGE_FAULT_IN_NONPAGED_AREA`) 

> I noticed application crashes from time to time and BSoDs with the message "A clock interrupt was not received on a secondary processor within the allocated time interval".
> I'm afraid isn't related to this issue but supposed is good to know.

It must be related to #576 .
I sent an email on the `xen-devel` mailing list to discuss a possible race condition in altp2m APIs, and see what the Xen devs think of the current implementation.

@skvl 
Thanks for confirming the application crashes also.
You might want to investigate them with `WinDBG` to understand the root cause, and if you get similar results like `invalid opcode` or `access violation` in locations where the context should have been valid.

**Comment by tklengyel** [comment]
@mtarral I did a fresh installation on my CI machine of Windows 10 1903 and I can't reproduce the bluescreens you are seeing. I did multiple injections, started and stopped DRAKVUF several times with all plugins enabled and also with just syscalls. While the performance is not ideal with all plugins enabled the system didn't bluescreen. 

One issue I noticed though is that it takes about ~30s for the VM to resume execution after it was restored from a save file. That's probably an unrelated issue.

**Comment by mtarral** [comment]
@tklengyel thank you for testing.

The Windows build I'm using is `1803` (March 2018)

I installed a build `1903`, professional edition, and tested again, to see if KPTI was more stable in recent releases.

Result is the following (4 VCPUs):
![win10-1903-fail](https://user-images.githubusercontent.com/49681011/57917239-93b33900-7894-11e9-87e9-3ea0b5d2c387.png)

This build is more stable than the `1803`, i managed to execute `52` tests before a crash.
That's why you had trouble to reproduce this issue, it's a race-condition that needs thorough testing and a test framework to be triggered.

I propose to share my test framework with the Drakvuf community, I'm sure we will benefit from it, to validate the plugins, and ensure that drakvuf doesn't have unwanted behavior in the guest.

@skvl I saw that you created `drakvuf-runner`:
https://github.com/skvl/drakvuf-runner

I'm sure it can be helpful for you too.

PS: I tested again with 1 VCPU to see if the guest was more resiliant, and it's the opposite !
the guest is in a "frozen" state after ~10 tests.
It never happened before.
